texar/torch/modules/tokenizers/pretrained_tokenizer_base.py

```diff
@@ -0,0 +1,451 @@
+        all_vocab_files_names = {
+            'added_tokens_file': ADDED_TOKENS_FILE,
+            'special_tokens_map_file': SPECIAL_TOKENS_MAP_FILE,
+            'config_file': CONFIG_FILE}
+
+        # If a path to a file was provided, get the parent directory
+        saved_directory = pretrained_model_path
+        if os.path.exists(saved_directory) and not os.path.isdir(
+                saved_directory):
+            saved_directory = os.path.dirname(saved_directory)
+
+        for file_id, file_name in all_vocab_files_names.items():
+            full_file_name = os.path.join(saved_directory, file_name)
+            if not os.path.exists(full_file_name):
+                print("Can't find file {}. We won't load it.".format(
+                    full_file_name))
+                full_file_name = None
+            vocab_files[file_id] = full_file_name
+
+        if all(full_file_name is None for full_file_name in
+               vocab_files.values()):
+            raise ValueError("Can't find tokenizer files in {}.".format(
+                saved_directory))
+
+        kwargs: Dict[str, Any] = {'pretrained_model_name': None}
+
+        added_tokens_file = vocab_files.pop('added_tokens_file', None)
+        special_tokens_map_file = vocab_files.pop(
+            'special_tokens_map_file', None)
+        tokenizer_config_file = vocab_files.pop('config_file', None)
+
+        for args_name, file_path in vocab_files.items():
+            if args_name not in kwargs:
+                kwargs[args_name] = file_path
+
+        if special_tokens_map_file is not None:
+            with open(special_tokens_map_file, encoding="utf-8") as f:
+                special_tokens_map = json.load(f)
+            for key, value in special_tokens_map.items():
+                if key not in kwargs:
+                    kwargs[key] = value
+
+        if tokenizer_config_file is not None:
+            with open(tokenizer_config_file, encoding="utf-8") as f:
+                tokenizer_config = json.load(f)
+            for key, value in tokenizer_config.items():
+                if key not in kwargs:
+                    kwargs[key] = value
+
+        if hparams is not None:
+            for key, value in hparams.items():
+                kwargs[key] = value
+
+        tokenizer = cls(hparams=kwargs)
+
+        # Add supplementary tokens.
+        if added_tokens_file is not None:
+            with open(added_tokens_file, encoding="utf-8") as f:
+                added_tok_encoder = json.load(f)
+            added_tok_decoder = {v: k for k, v in added_tok_encoder.items()}
+            tokenizer.added_tokens_encoder.update(added_tok_encoder)
+            tokenizer.added_tokens_decoder.update(added_tok_decoder)
+
+        return tokenizer
+
+    def save(self, save_directory: str) -> Tuple[str]:
+        r"""Save the tokenizer vocabulary files (with added tokens) and the
+        `special-tokens-to-class-attributes-mapping` to a directory, so that it
```

Comment:
what's `special-tokens-to-class-attributes-mapping`? Make it clearer in the docstring